File: figures/08-graphs/snippet-bellman-ford.py
def bellman_ford(g, s):
    n = g.vertices()
    dist = [INF] * n
    parent = [-1] * n
    dist[s] = 0
    arcs = g.arcs()
    for _ in range(1, n):
        changed = False
        for u, v, w in arcs:
            if dist[u] != INF and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                changed = True
        if not changed:
            break
    negative_cycle = any(
        dist[u] != INF and dist[u] + w < dist[v] for u, v, w in arcs
    )
    return dist, parent, negative_cycle

File: figures/08-graphs/snippet-bellman-ford.java
    public static Result bellmanFord(Graph g, int s) {
        int n = g.vertices(); long[] d = new long[n]; int[] p = new int[n];
        Arrays.fill(d, INF); Arrays.fill(p, -1); d[s] = 0;
        List<long[]> arcs = g.arcs();
        for (int i = 1; i < n; i++) {
            boolean changed = false;
            for (long[] e : arcs) if (d[(int)e[0]] != INF) {
                long cand = safeAdd(d[(int)e[0]], e[2]);
                if (cand < d[(int)e[1]]) {
                    d[(int)e[1]] = cand; p[(int)e[1]] = (int)e[0]; changed = true;
                }
            }
            if (!changed) break;
        }
        for (long[] e : arcs) if (d[(int)e[0]] != INF &&
                safeAdd(d[(int)e[0]], e[2]) < d[(int)e[1]])
            return new Result(d, p, true);
        return new Result(d, p, false);
    }
